man/stack/sp_stack_remove.3
.\"  Staple - A general-purpose data structure library in pure C89.
.\"  Copyright (C) 2021  Randoragon
.\"
.\"  This library is free software; you can redistribute it and/or
.\"  modify it under the terms of the GNU Lesser General Public
.\"  License as published by the Free Software Foundation;
.\"  version 2.1 of the License.
.\"
.\"  This library is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\"  Lesser General Public License for more details.
.\"
.\"  You should have received a copy of the GNU Lesser General Public
.\"  License along with this library; if not, write to the Free Software
.\"  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
.\"--------------------------------------------------------------------------------
.TH SP_STACK_REMOVE 3 DATE "libstaple-VERSION"
.SH NAME
sp_stack_remove,
sp_stack_removec,
sp_stack_removes,
sp_stack_removei,
sp_stack_removel,
sp_stack_removesc,
sp_stack_removeuc,
sp_stack_removeus,
sp_stack_removeui,
sp_stack_removeul,
sp_stack_removef,
sp_stack_removed,
sp_stack_removeld,
sp_stack_removell,
sp_stack_removeull,
sp_stack_removeu8,
sp_stack_removeu16,
sp_stack_removeu32,
sp_stack_removeu64,
sp_stack_removei8,
sp_stack_removei16,
sp_stack_removei32,
sp_stack_removei64,
sp_stack_removestr
\- remove an element from any position of a stack
.SH SYNOPSIS
.ad l
#include <staple.h>
.sp
int
.BR sp_stack_remove "(struct sp_stack"
.RI * stack ,
size_t
.IR idx ,
int
.RI (* dtor )(void*))
.br
char
.BR sp_stack_removec "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
short
.BR sp_stack_removes "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
int
.BR sp_stack_removei "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
long
.BR sp_stack_removel "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
signed char
.BR sp_stack_removesc "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
unsigned char
.BR sp_stack_removeuc "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
unsigned short
.BR sp_stack_removeus "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
unsigned int
.BR sp_stack_removeui "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
unsigned long
.BR sp_stack_removeul "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
float
.BR sp_stack_removef "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
double
.BR sp_stack_removed "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
long double
.BR sp_stack_removeld "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
long long
.BR sp_stack_removell "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
unsigned long long
.BR sp_stack_removeull "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
uint8_t
.BR sp_stack_removeu8 "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
uint16_t
.BR sp_stack_removeu16 "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
uint32_t
.BR sp_stack_removeu32 "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
uint64_t
.BR sp_stack_removeu64 "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
int8_t
.BR sp_stack_removei8 "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
int16_t
.BR sp_stack_removei16 "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
int32_t
.BR sp_stack_removei32 "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
int64_t
.BR sp_stack_removei64 "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.br
char
.RB * sp_stack_removestr "(struct sp_stack"
.RI * stack ,
size_t
.IR idx )
.sp
Link with \fI-lstaple\fP.
.ad
.SH DESCRIPTION
The functions in the
.BR sp_stack_remove ()
family are used to remove elements from any position of a stack. The order of
remaining elements shall be preserved.
.P
.I stack
is the stack that shall be removed from.
.P
.I idx
is the index of the element that shall be removed. Indices start from 0 and
extend downwards \(em 0 is the top element, 1 is the element below it, and so
on.
.br
Removing position
.BR n " requires " n
elements to be moved in memory. Removing the 0th element has the same effect as
.BR sp_stack_pop (3).
.br
.I idx
must be smaller than
.IR stack->size .
.SS Generic Form
.I dtor
is a pointer to a destructor function that shall be invoked on the removed
element. This mechanism is useful if individual stack elements own memory
that should be freed (see EXAMPLES). If no such destructor function is needed,
.B NULL
can be passed to skip this functionality.
.SH RETURN VALUE
.SS Generic Form
If successful, the
.BR sp_stack_remove ()
function shall return 0. Otherwise, it shall return a code indicating the
error.
.SS Suffixed Form
The suffixed
.BR sp_stack_remove ()
functions shall return the value of the removed element.
.P
The
.BR sp_stack_removestr ()
function returns a pointer to the memory block containing the removed string.
The user must manually
.BR free (3)
the block, otherwise the memory will leak!
.SH ERRORS
.SS Generic Form
The
.BR sp_stack_remove ()
function shall fail if:
.IP \fBSP_EINVAL\fP 1.5i
.I stack
is a null pointer (debug mode only).
.IP \fBSP_EINDEX\fP 1.5i
.I idx
out of range (debug mode only).
.IP \fBSP_ECALLBK\fP 1.5i
.I dtor
returned non-0.
.SS Suffixed Form
The suffixed
.BR sp_stack_remove ()
functions shall fail if:
.IP \fB0\fP 1.5i
.I stack
is a null pointer (debug mode only).
.IP \fB0\fP 1.5i
.I idx
out of range (debug mode only).
.IP \fB0\fP 1.5i
.IR stack->elem_size
does not match the size of the function's return type (debug mode only).
.P
(the
.BR sp_stack_removestr ()
function returns
.B NULL
instead of 0)
.SH NOTES
In debug mode, functions of the suffixed form (except
.BR sp_stack_removestr ())
do not provide an unambiguous error-detecting interface with return codes.
Returning 0 should be considered more so a fallback mechanism, since it doesn't
imply that an error has occurred.
.P
This means there is a guaranteed runtime uncertainty whenever one of these
functions in debug mode returns 0. However, the functions shall still print
error messages when appropriate, so failures can be told apart from successes by
the presence or absence of those (quiet mode needs to be disabled).
.P
At first, this ambiguity might seem strange and unacceptable. However, it is
justified in two ways:
.IP (1)
It is very convenient to have these functions return the value of the element
directly.
.sp -1
.IP (2)
The only errors that can occur are such resulting from library misuse by the
end-user, and not caused by uncontrollable factors such as insufficient memory,
etc. Program logic itself should guarantee safety from these errors.
.P
The generic
.BR sp_stack_remove ()
function does not return the removed element. To work around this limitation,
one should use
.BR sp_stack_peek ()
or
.BR sp_stack_get ()
to manually copy the element prior to its removal:
.IP
.ad l
.nf
/* If element type is known */
elem_t elem = *(elem_t*)sp_stack_get(stack, n);

/* If only element size is known */
void *elem = malloc(stack->elem_size);
memcpy(elem, sp_stack_get(stack, n), stack->elem_size);
.fi
.ad
.SH CONFORMING TO
See
.BR libstaple (7).
.SH EXAMPLES
This is an example of how to use the
.I dtor
function:
.IP
.ad l
.nf
struct example {
        void *ptr;
};

int example_dtor(void *elem)
{
        free(((struct example*)elem)->ptr);
        return 0;
}

int main()
{
        struct sp_stack *s;
        struct example data;
        s = sp_stack_create(sizeof(struct example), 1);
        data.ptr = malloc(32);
        sp_stack_push(s, &data);
        sp_stack_remove(s, 0, &example_dtor);
        /* The above function call is equivalent to:
                example_dtor(sp_stack_peek(s))
                sp_stack_remove(s, 0, NULL);
        */
        return 0;
}
.fi
.ad
.SH SEE ALSO
.ad l
.BR libstaple (7),
.BR sp_stack (7),
.BR sp_stack_create (3),
.BR sp_stack_destroy (3),
.BR sp_stack_clear (3),
.BR sp_stack_push (3),
.BR sp_stack_peek (3),
.BR sp_stack_pop (3),
.BR sp_stack_insert (3),
.BR sp_stack_qinsert (3),
.BR sp_stack_qremove (3),
.BR sp_stack_get (3),
.BR sp_stack_set (3),
.BR sp_stack_eq (3),
.BR sp_stack_copy (3),
.BR sp_stack_map (3),
.BR sp_stack_print (3),
.BR free (3)


src/stack/sp_stack_remove.c
/*  Staple - A general-purpose data structure library in pure C89.
 *  Copyright (C) 2021  Randoragon
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation;
 *  version 2.1 of the License.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "../sp_stack.h"
#include "../internal.h"
#include "../sp_errcodes.h"
#include <string.h>

int sp_stack_remove(struct sp_stack *stack, size_t idx, int (*dtor)(void*))
{
	char *p;
	int err;
#ifdef STAPLE_DEBUG
	if (stack == NULL) {
		error(("stack is NULL"));
		return SP_EINVAL;
	}
	if (idx >= stack->size) {
		error(("index out of range"));
		return SP_EINDEX;
	}
#endif
	p = (char*)stack->data + (stack->size - 1 - idx) * stack->elem_size;
	if (dtor != NULL && (err = dtor(p))) {
		error(("callback function dtor returned %d (non-0)", err));
		return SP_ECALLBK;
	}
	memmove(p, p + stack->elem_size, idx * stack->elem_size);
	--stack->size;
	return 0;
}
